File: ruby/leetcode/FindFirstAndLastPositionOfElementInSortedArray_34.rb
# 34 Find First and Last Position of Element in Sorted Array

# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def search_range(nums, target)
  s = nums.index(target)
  return [-1, -1] unless s

  nums = nums.reverse
  e = nums.length - 1 - nums.index(target)

  [s, e]
end

nums = [5,7,7,8,8,10]
target = 8
p search_range(nums, target)

File: java/problem_solving/src/leetcode/FindFirstAndLastPositionOfElementInSortedArray_34.java
package leetcode;

// #34 - Find First And Last Position Of Element In Sorted Array
import java.util.Arrays;

public class FindFirstAndLastPositionOfElementInSortedArray_34 {
	public static void main(String[] args) {
		int[] nums = {5, 7, 7, 8, 8, 10};
		int target = 6;

		// int[] nums = {1};
		// int target = 1;

		System.out.println("searchRange(nums,targer) = " + Arrays.toString(searchRange(nums, target)));
	}

	public static int[] searchRange(int[] nums, int target) {
		int l = 0, r = nums.length - 1;
		int[] result = {-1, -1};
		while (l <= r) {
			int mid = (l + r) / 2;
			int t = nums[mid];
			if (t < target) {
				l = mid + 1;
			} else {
				if (t == target) result[0] = mid;
				r = mid - 1;
			}
		}

		l = 0;
		r = nums.length - 1;
		while (l <= r) {
			int mid = (l + r) / 2;
			int t = nums[mid];
			if (t > target) {
				r = mid - 1;
			} else {
				if (t == target) result[1] = mid;
				l = mid + 1;
			}
		}
		return result;
	}
}
